### PR TITLE
Pilot: Allow '~' in ipa path

### LINKED ISSUE #9480: [Pilot] Cannot reference home directory in "ipa" parameter
### New Issue Checklist

- [X] Updated fastlane to the latest version
- [X] I have read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)

### Issue Description

Passing a path including a reference to the user's home directory fails in Pilot.

```
    pilot(
      # other args...
      ipa: "~/some/path/to/app.ipa"
    )
```

### Captured Output

Command Used: `foo`
<details><summary>Output/Log</summary>

```

INFO [2017-06-14 16:32:05.48]: Driving the lane 'ios foo' 🚀
INFO [2017-06-14 16:32:05.48]: -------------------
INFO [2017-06-14 16:32:05.48]: --- Step: pilot ---
INFO [2017-06-14 16:32:05.48]: -------------------
ERROR [2017-06-14 16:32:05.51]: Error setting value '~/builds/iOS/core_app/prod/MyApp_243_1.ipa' for option 'ipa'
ERROR [2017-06-14 16:32:05.51]: You passed invalid parameters to 'pilot'.
ERROR [2017-06-14 16:32:05.51]: Check out the error below and available options by running `fastlane action pilot`
WARN [2017-06-14 16:32:05.90]: Lane Context:
INFO [2017-06-14 16:32:05.90]: {:DEFAULT_PLATFORM=>:ios, :PLATFORM_NAME=>:ios, :LANE_NAME=>"ios foo"}
ERROR [2017-06-14 16:32:05.90]: Could not find ipa file at path '~/builds/iOS/core_app/prod/MyApp_243_1.ipa'

+------+---------------------+-------------+
|             fastlane summary             |
+------+---------------------+-------------+
| Step | Action              | Time (in s) |
+------+---------------------+-------------+
| 1    | Verifying required  | 0           |
|      | fastlane version    |             |
| 2    | xcversion           | 4           |
| 3    | xcode_select        | 0           |
| 4    | default_platform    | 0           |
| 💥   | pilot               | 0           |
+------+---------------------+-------------+

ERROR [2017-06-14 16:32:05.93]: fastlane finished with errors


```

</details>

<details><summary>✅ fastlane environment ✅</summary>

### Stack

| Key                         | Value                                       |
| --------------------------- | ------------------------------------------- |
| OS                          | 10.12.4                                     |
| Ruby                        | 2.2.4                                       |
| Bundler?                    | false                                       |
| Git                         | git version 2.11.0 (Apple Git-81)           |
| Installation Source         | ~/.rbenv/versions/2.2.4/bin/fastlane        |
| Host                        | Mac OS X 10.12.4 (16E195)                   |
| Ruby Lib Dir                | ~/.rbenv/versions/2.2.4/lib                 |
| OpenSSL Version             | OpenSSL 1.0.2k  26 Jan 2017                 |
| Is contained                | false                                       |
| Is homebrew                 | false                                       |
| Is installed via Fabric.app | false                                       |
| Xcode Path                  | /Applications/Xcode.app/Contents/Developer/ |
| Xcode Version               | 8.3.2                                       |

### System Locale

| Variable | Value       |   |
| -------- | ----------- | - |
| LANG     | en_US.UTF-8 | ✅ |
| LC_ALL   |             |   |
| LANGUAGE |             |   |

### PR DESCRIPTION
**Pull request by nickffox**
### Checklist
- [X] I've run `bundle exec rspec` from the root directory to see all new and existing tests pass
- [X] I've followed the _fastlane_ code style and run `bundle exec rubocop -a` to ensure the code style is valid
- [X] I've read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [X] I've updated the documentation if necessary.

### Motivation and Context

Fixes #9480

### CODE REVIEW

File: fastlane_core/lib/fastlane_core/ipa_file_analyser.rb [hunk 1]
```diff
@@ -19,6 +19,7 @@ def self.fetch_app_version(path)
 
     # Fetches the app platform from the given ipa file.
     def self.fetch_app_platform(path)
+      path = File.expand_path(path)
```
Comment (ohayon) [inline fastlane_core/lib/fastlane_core/ipa_file_analyser.rb]: Should this be happening in the methods above too? Or maybe at the call site for all of these methods?

**Review by ohayon — COMMENTED** [review]
Thanks for the PR @NicholasFFox! This is definitely a great change! I just left one comment about the methods in the `IpaFileAnalyser` class. Let me know what you think? 🐙

File: fastlane_core/lib/fastlane_core/ipa_file_analyser.rb (hunk 1 above)
Comment (NicholasFFox) [inline fastlane_core/lib/fastlane_core/ipa_file_analyser.rb]: @ohayon I _really_ wanted to make the change in `fastlane_core/lib/fastlane_core/configuration/config_item.rb` by changing `auto_convert_value` to do this: 
```
    def auto_convert_value(value)
      return nil if value.nil?

      if data_type == Array
        return value.split(',') if value.kind_of?(String)
      elsif data_type == Integer
        return value.to_i if value.to_i.to_s == value.to_s
      elsif data_type == Float
        return value.to_f if value.to_f.to_s == value.to_s
      elsif data_type == String && allow_shell_conversion
        return value.map(&:to_s).map(&:shellescape).join(' ') if value.kind_of?(Array)
        return value.map { |k, v| "#{k.to_s.shellescape}=#{v.shellescape}" }.join(' ') if value.kind_of?(Hash)
      elsif data_type == String && value.include?("~/") # <- new case here
        return File.expand_path(value)
      else
        # Special treatment if the user specified true, false or YES, NO
        # There is no boolean type, so we just do it here
        if %w(YES yes true TRUE).include?(value)
          return true
        elsif %w(NO no false FALSE).include?(value)
          return false
        end
      end

      return value # fallback to not doing anything
    end
```
That fix broke a number of tests though that had `~/` in various paths, so I opted for the less global fix.

**Comment by NicholasFFox** [comment]
@ohayon updated all methods in IpaFileAnalyser to expand paths.

File: pilot/lib/pilot/build_manager.rb [hunk 2]
```diff
@@ -14,8 +14,9 @@ def upload(options)
       dir = Dir.mktmpdir
 
       platform = fetch_app_platform
+      ipa_path = File.expand_path(config[:ipa])
```
Comment (ohayon) [inline pilot/lib/pilot/build_manager.rb]: Any chance you would want to move this into the `generate` method on `IpaUploadPackageBuilder` to let that method handle the expansion?

**Review by ohayon — CHANGES_REQUESTED** [review]
Hey @NicholasFFox, thanks for making those changes!  And sorry I am asking one more thing, but I'm curious to hear your thoughts on the one comment I left. 💯

**Comment by NicholasFFox** [comment]
@ohayon completely agree with the comment and made the requested change 👍

**Comment by KrauseFx** [comment]
Hey everyone, randomly jumping in here. I agree with @ohayon, this ends up with lots of duplicate code. An easy solution for that would be to do it once in `gym/manager.rb` right after printing the tables (as we don't necessarily want to display a really long path in the summary table) by setting it once:

```ruby
Gym.config[:ipa_path] = File.expand_path(Gym.config[:ipa_path])
```

or similar :)

**Comment by NicholasFFox** [comment]
@KrauseFx @ohayon - I reverted my previous changes (with the exception of the ipa parameter's verify block), and added a `create_config` method in Pilot's CommandsGenerator. This offers a central place to perform similar operations on Pilot parameters in the future. Does this address the previous concerns?

**Review by KrauseFx — APPROVED** [review]
Looks 💯

**Comment by NicholasFFox** [comment]
@ohayon anything else that I need to do here?

**Review by ohayon — APPROVED** [review]

**Comment by KrauseFx** [comment]
Thanks for reminding us @NicholasFFox, this is merged now and will be in the next release 👍 Thanks for contributing.

**Pull request merged**